salt/modules/transactional_update.py

```diff
@@ -985,13 +985,30 @@ def call(function, *args, **kwargs):
             else:
                 return local
         except ValueError:
-            return {"result": False, "retcode": 1, "comment": ret_stdout}
+            local = {"result": False, "retcode": 1, "comment": ret_stdout}
+            return local
     finally:
         # Check if reboot is needed
-        if activate_transaction and pending_transaction():
+        if (activate_transaction and pending_transaction()) or _user_specified_reboot(
+            local
+        ):
             reboot()
 
 
+def _user_specified_reboot(local):
+    if not isinstance(local, dict):
+        # Skip if execution is not state/highstate
+        return False
+
+    explicit_reboot_cmds = ["reboot", "init 6", "shutdown -r", "shutdown --reboot"]
+    names = []
+    for _, value in local.items():
+        if isinstance(value, dict) and "name" in value:
```

Comment:
Does this match all kinds of state functions? If so, we risk rebooting when it wasn't actually requested. I think we should prefer false-negatives over false-positives. Rebooting when it's not asked for is disruptive while a missing reboot can be worked around.

Reply:
I've modified the actual matching below since this comment; so while we do iterate over all state functions, the function is matched with exact match only. Right now, we only match names that equal to `reboot` or `system.reboot` (from `module.include`). Is that sufficient to alleviate your concern, or do you think we should also match the key? (e.g. `cmd` or `module` in `key`)

Reply:
> Side note: the following sls seems not to start a transaction, so this change doesn't affect it.

Is that really the case? Without this patch, the condition to decide if a reboot should be called requires a pending transaction. With this patch, it's either pending transaction + activate_transaction=True or we find `name: reboot`. Am I missing something?

> The same applies to state.test - it also doesn't run in a transaction.
Right, that's the default. We might want to still rule that out (as you've done anyway), the configuration what runs inside/outside a transaction can be configured.

Reply:
No indeed, you were right. When I was testing it, I think I mistakenly turned off the t-u executor. Upon re-checking my results, I realized I was wrong, and added a further check